apps/framework-cli/src/cli/routines/query.rs

```diff
@@ -0,0 +1,183 @@
+) -> Result<RoutineSuccess, RoutineFailure> {
+    let sql_query = get_sql_input(sql, file)?;
+    info!("Executing SQL: {}", sql_query);
+
+    // Get ClickHouse connection pool
+    let pool = get_pool(&project.clickhouse_config);
+
+    let mut client = pool.get_handle().await.map_err(|_| {
+        RoutineFailure::error(Message::new(
+            "Failed".to_string(),
+            "Error connecting to storage".to_string(),
+        ))
+    })?;
+
+    let redis_client = setup_redis_client(project.clone()).await.map_err(|e| {
+        RoutineFailure::error(Message {
+            action: "Query".to_string(),
+            details: format!("Failed to setup redis client: {e:?}"),
+        })
+    })?;
+
+    // Validate that infrastructure state exists and is accessible.
+    // The value is not used further, but we fail early if it cannot be loaded.
+    let _infra = InfrastructureMap::load_from_redis(&redis_client)
+        .await
+        .map_err(|_| {
+            RoutineFailure::error(Message::new(
+                "Failed".to_string(),
+                "Error retrieving current state".to_string(),
+            ))
+        })?
+        .ok_or_else(|| {
+            RoutineFailure::error(Message::new(
+                "Failed".to_string(),
+                "No state found".to_string(),
+            ))
+        })?;
+
+    // Execute query and stream results
+    let mut stream = client.query(&sql_query).stream();
+
+    let mut success_count = 0;
+    let mut enum_mappings: Option<Vec<Option<Vec<&str>>>> = None;
+
+    while let Some(row_result) = stream.next().await {
+        match row_result {
+            Ok(row) => {
+                // Create enum mappings on first row (one None entry per column)
+                if enum_mappings.is_none() {
+                    enum_mappings = Some(vec![None; row.len()]);
+                }
+
+                // Reuse peek's row_to_json with enum mappings
+                let value = row_to_json(&row, enum_mappings.as_ref().unwrap()).map_err(|e| {
+                    RoutineFailure::new(
+                        Message::new(
+                            "Query".to_string(),
+                            "Failed to convert row to JSON".to_string(),
+                        ),
+                        e,
+                    )
+                })?;
+
+                let json = serde_json::to_string(&value).map_err(|e| {
+                    RoutineFailure::new(
+                        Message::new(
+                            "Query".to_string(),
+                            "Failed to serialize result".to_string(),
+                        ),
+                        e,
+                    )
+                })?;
+
+                println!("{}", json);
+                info!("{}", json);
+                success_count += 1;
+
+                // Check limit to avoid unbounded queries
+                if success_count >= limit {
+                    info!("Reached limit of {} rows", limit);
```

Comment:
we do it this way because the rust client does not support max_count_row client options. we'd have to switch to the http client to get those options.